### ISSUE TITLE
summary stats support?

### ISSUE THREAD

**Issue by KnutJaegersberg** [xiaodaigh/disk.frame#206]
Hi,
do you plan direct summary stats support like in juliadb / onlinestats? you can go far with the two stage groupby manually, yet this feels like very basic functionality which would be nice to work out of the box.

**Comment by xiaodaigh** [comment]
Yeah definitely! There is a project to redo the group by so that many common functions like mean can be done in one stage with no change of code from dplyr or data.table. See https://github.com/xiaodaigh/disk.frame/projects/2

However this is an interesting technical project and might be quite involved. So it might take a while given I try to work on OSS on Sunday mornings only. 

The implementation will basically define a few functions using a two stage approach. I believe I saw something in Dask about how to define your own aggregation functions, so probably study that a bit and make it happen.

**Comment by xiaodaigh** [comment]
related to #128 and #128

**Comment by xiaodaigh** [comment]
Tracking progress here: https://github.com/xiaodaigh/disk.frame/pull/219

**Comment by xiaodaigh** [comment]
@KnutJaegersberg one -stage group- by is available now in the `chunk-group-by` branch, see example code at end.

Currently, the following functions are supported

`mean, sum, sd, var, max, min, median, length`.

Please help test it if you can. When your own examples, I am working on a mechanism to allow the user to define their own group-by functions, so they are not limited by the lists provided by {disk.frame}.

The plan is for this experimental feature to be in the 0.2.2 release, which will show some deprecation messages.  Then in 0.3.0 all deprecate messages will be removed and this group-by framework will be the default.

```r
remotes::install_github("xiaodaigh/disk.frame", ref = "chunk-group-by")

library(disk.frame)
setup_disk.frame()

result_from_disk.frame = iris %>% 
  as.disk.frame %>% 
  group_by(Species) %>% 
  summarize(
    mean(Petal.Length), 
    sumx = sum(Petal.Length/Sepal.Width), 
    sd(Sepal.Width/ Petal.Length), 
    var(Sepal.Width/ Sepal.Width), 
    l = length(Sepal.Width/ Sepal.Width + 2),
    max(Sepal.Width), 
    min(Sepal.Width), 
    median(Sepal.Width)
    ) %>% 
  collect


result_from_dplyr = iris %>% 
  group_by(Species) %>% 
  summarize(
    mean(Petal.Length), 
    sumx = sum(Petal.Length/Sepal.Width), 
    sd(Sepal.Width/ Petal.Length), 
    var(Sepal.Width/ Sepal.Width), 
    l = length(Sepal.Width/ Sepal.Width + 2),
    max(Sepal.Width), 
    min(Sepal.Width), 
    median(Sepal.Width)
  )

# tests to ensure results are all the same
setequal(names(result_from_dplyr), names(result_from_disk.frame))
nrow(result_from_dplyr) == nrow(result_from_disk.frame)

for(n in names(result_from_dplyr)) {
  stopifnot(
    all(result_from_dplyr[, n] == result_from_disk.frame[, n]) || 
    all(abs(result_from_dplyr[, n] - result_from_disk.frame[, n]) < 0.0001)
  )
}
 
```

**Comment by xiaodaigh** [comment]
closed by #223 #224
